fips202/keccakf1600.c

```diff
@@ -1,321 +1,938 @@
+    temp = (temp0 ^ (temp0 >>  8)) & 0x0000FF00UL;  temp0 = temp0 ^ temp ^ (temp <<  8); \
+    temp1 = (high); \
+    temp = (temp1 ^ (temp1 >>  1)) & 0x22222222UL;  temp1 = temp1 ^ temp ^ (temp <<  1); \
+    temp = (temp1 ^ (temp1 >>  2)) & 0x0C0C0C0CUL;  temp1 = temp1 ^ temp ^ (temp <<  2); \
+    temp = (temp1 ^ (temp1 >>  4)) & 0x00F000F0UL;  temp1 = temp1 ^ temp ^ (temp <<  4); \
+    temp = (temp1 ^ (temp1 >>  8)) & 0x0000FF00UL;  temp1 = temp1 ^ temp ^ (temp <<  8);
+
+#define toBitInterleavingAndXOR(low, high, even, odd, temp, temp0, temp1) \
+    prepareToBitInterleaving(low, high, temp, temp0, temp1) \
+    even ^= (temp0 & 0x0000FFFF) | (temp1 << 16); \
+    odd ^= (temp0 >> 16) | (temp1 & 0xFFFF0000);
+
+#define toBitInterleavingAndAND(low, high, even, odd, temp, temp0, temp1) \
+    prepareToBitInterleaving(low, high, temp, temp0, temp1) \
+    even &= (temp0 & 0x0000FFFF) | (temp1 << 16); \
+    odd &= (temp0 >> 16) | (temp1 & 0xFFFF0000);
+
+#define toBitInterleavingAndSet(low, high, even, odd, temp, temp0, temp1) \
+    prepareToBitInterleaving(low, high, temp, temp0, temp1) \
+    even = (temp0 & 0x0000FFFF) | (temp1 << 16); \
+    odd = (temp0 >> 16) | (temp1 & 0xFFFF0000);
+
+/* Credit to Henry S. Warren, Hacker's Delight, Addison-Wesley, 2002 */
+#define prepareFromBitInterleaving(even, odd, temp, temp0, temp1) \
+    temp0 = (even); \
+    temp1 = (odd); \
+    temp = (temp0 & 0x0000FFFF) | (temp1 << 16); \
+    temp1 = (temp0 >> 16) | (temp1 & 0xFFFF0000); \
+    temp0 = temp; \
+    temp = (temp0 ^ (temp0 >>  8)) & 0x0000FF00UL;  temp0 = temp0 ^ temp ^ (temp <<  8); \
+    temp = (temp0 ^ (temp0 >>  4)) & 0x00F000F0UL;  temp0 = temp0 ^ temp ^ (temp <<  4); \
+    temp = (temp0 ^ (temp0 >>  2)) & 0x0C0C0C0CUL;  temp0 = temp0 ^ temp ^ (temp <<  2); \
+    temp = (temp0 ^ (temp0 >>  1)) & 0x22222222UL;  temp0 = temp0 ^ temp ^ (temp <<  1); \
+    temp = (temp1 ^ (temp1 >>  8)) & 0x0000FF00UL;  temp1 = temp1 ^ temp ^ (temp <<  8); \
+    temp = (temp1 ^ (temp1 >>  4)) & 0x00F000F0UL;  temp1 = temp1 ^ temp ^ (temp <<  4); \
+    temp = (temp1 ^ (temp1 >>  2)) & 0x0C0C0C0CUL;  temp1 = temp1 ^ temp ^ (temp <<  2); \
+    temp = (temp1 ^ (temp1 >>  1)) & 0x22222222UL;  temp1 = temp1 ^ temp ^ (temp <<  1);
+
+#define fromBitInterleaving(even, odd, low, high, temp, temp0, temp1) \
+    prepareFromBitInterleaving(even, odd, temp, temp0, temp1) \
+    low = temp0; \
+    high = temp1;
+
+typedef struct {
+    uint32_t A[50];
+} KeccakP1600_plain32_state;
+
+static void KeccakP1600_AddBytesInLane(KeccakP1600_plain32_state *state, unsigned int lanePosition, const unsigned char *data, unsigned int offset, unsigned int length) {
+    uint8_t laneAsBytes[8];
+    uint32_t low, high;
+    uint32_t temp, temp0, temp1;
+    uint32_t *stateAsHalfLanes = state->A;
+
+    memset(laneAsBytes, 0, 8);
+    memcpy(laneAsBytes + offset, data, length);
+    #if (PLATFORM_BYTE_ORDER == IS_LITTLE_ENDIAN)
```

Comment:
Where is this `PLATFORM_BYTE_ORDER` configured ?

Reply:
Good catch. I added that.